src/Compilers/Core/Portable/Text/StringText.cs

```diff
@@ -90,14 +90,35 @@ public override void CopyTo(int sourceIndex, char[] destination, int destination
 
         public override void Write(TextWriter textWriter, TextSpan span, CancellationToken cancellationToken = default(CancellationToken))
         {
-            if (span.Start == 0 && span.End == this.Length)
+            CheckSubSpan(span);
+
+            if (span.Length == 0)
+                return;
+
+            if (span.Start == 0 && span.Length == this.Length)
             {
                 textWriter.Write(this.Source);
+                return;
+            }
+
+#if NETCOREAPP
+            var @delegate = (TextWriterWriteReadOnlySpanDelegate)textWriter.Write;
+            if (@delegate.Method.DeclaringType != typeof(TextWriter))
+            {
+                // Use the Write(ReadOnlySpan<char>) overload only if the method is overriden, because if it's not, the base
+                // TextWriter implementation needs to delegate to the array overload and could potentially allocate a really
+                // large array, which would be slower than the buffering base implementation.
+                textWriter.Write(this.Source.AsSpan(span.Start, span.Length));
             }
             else
             {
                 base.Write(textWriter, span, cancellationToken);
             }
+#else
+            base.Write(textWriter, span, cancellationToken);
+#endif
         }
+
+        private delegate void TextWriterWriteReadOnlySpanDelegate(ReadOnlySpan<char> buffer);
```

Comment:
It's a bit of a shame that I can't somehow specify the overload using parameter types when referring to a method. If I could, the "method group" could have a natural delegate type and I wouldn't have to declare one.